compiler/pipes/analyze-performance.cpp

```diff
@@ -0,0 +1,636 @@
+    case op_var:
+      if (const auto *constexpr_str = GenTree::get_constexpr_string(expr)) {
+        std::string raw_str{constexpr_str->c_str(), std::min(constexpr_str->size(), size_t{36})};
+        if (raw_str.size() < constexpr_str->size()) {
+          raw_str.append("<...>");
+        }
+        std::replace_if(raw_str.begin(), raw_str.end(),
+                        [](char c) {
+                          return iscntrl(c) || vk::any_of_equal(c, '"', '\'', '\\');
+                        }, '?');
+        raw_str.erase(std::unique(raw_str.begin(), raw_str.end(),
+                                  [](char lhs, char rhs) { return vk::any_of_equal(lhs, '?', ' ') && lhs == rhs; }),
+                      raw_str.end());
+        return "'" + vk::replace_all(raw_str, "?", "<...>") + "'";
+      }
+      return (full ? "variable " : "") + TermStringFormat::paint_green(expr.as<op_var>()->var_id->get_human_readable_name());
+    case op_instance_prop:
+      return (full ? "variable " : "") + TermStringFormat::paint_green(expr.as<op_instance_prop>()->var_id->get_human_readable_name());
+    case op_func_call:
+      return (full ? "function call " : "") + TermStringFormat::paint_green(expr.as<op_func_call>()->func_id->get_human_readable_name(false)) +
+             "(" + vk::join(expr.as<op_func_call>()->args(), ", ", get_description_for_help_impl<false>) + ")";
+    case op_index: {
+      std::string key;
+      auto op_index_vertex = expr.as<op_index>();
+      if (op_index_vertex->has_key()) {
+        key = get_description_for_help_impl<false>(op_index_vertex->key());
+      }
+      return (full ? "array element " : "") + get_description_for_help_impl<false>(op_index_vertex->array()) + "[" + key + "]";
+    }
+    case op_string_build:
+      return join_string_build_elements(expr.as<op_string_build>()->args());
+    case op_true:
+      return "true";
+    case op_false:
+      return "false";
+    case op_null:
+      return "null";
+    case op_int_const:
+    case op_float_const:
+      return expr.as<meta_op_num>()->str_val;
+    default:
+      // TODO add more cases
+      return full ? "expresion <...>" : "<...>";
+  }
+}
+
+std::string get_description_for_help(VertexPtr expr) noexcept {
+  return get_description_for_help_impl<true>(expr);
+}
+
+bool is_same_var_expression(VertexPtr lhs, VertexPtr rhs) noexcept {
+  if (!lhs || !rhs) {
+    return false;
+  }
+
+  lhs = remove_conv_wrap(lhs);
+  rhs = remove_conv_wrap(rhs);
+  if (lhs->type() != rhs->type()) {
+    return false;
+  }
+
+  switch (lhs->type()) {
+    case op_var:
+      if (lhs.as<op_var>()->var_id != rhs.as<op_var>()->var_id) {
+        return false;
+      }
+      break;
+    case op_instance_prop:
+      if (lhs.as<op_instance_prop>()->var_id != rhs.as<op_instance_prop>()->var_id) {
+        return false;
+      }
+      break;
+    case op_int_const:
+    case op_float_const:
+      if (lhs.as<meta_op_num>()->str_val != rhs.as<meta_op_num>()->str_val) {
```

Comment:
'0' != '0x0' in this case, I amn't sure if it matters

Reply:
I suppose we can leave as is for now.